// @fusionstrings/swiss-eph/esm/generated/api.js
// Auto-generated by scripts/codegen.ts
// Do not edit manually.
export const Constants = {
    SE_AUNIT_TO_KM: 149597870.700,
    SE_AUNIT_TO_LIGHTYEAR: 1.0 / 63241.07708427,
    SE_AUNIT_TO_PARSEC: 1.0 / 206264.8062471,
    SE_JUL_CAL: 0,
    SE_GREG_CAL: 1,
    SE_ECL_NUT: -1,
    SE_SUN: 0,
    SE_MOON: 1,
    SE_MERCURY: 2,
    SE_VENUS: 3,
    SE_MARS: 4,
    SE_JUPITER: 5,
    SE_SATURN: 6,
    SE_URANUS: 7,
    SE_NEPTUNE: 8,
    SE_PLUTO: 9,
    SE_MEAN_NODE: 10,
    SE_TRUE_NODE: 11,
    SE_MEAN_APOG: 12,
    SE_OSCU_APOG: 13,
    SE_EARTH: 14,
    SE_CHIRON: 15,
    SE_PHOLUS: 16,
    SE_CERES: 17,
    SE_PALLAS: 18,
    SE_JUNO: 19,
    SE_VESTA: 20,
    SE_INTP_APOG: 21,
    SE_INTP_PERG: 22,
    SE_NPLANETS: 23,
    SE_PLMOON_OFFSET: 9000,
    SE_AST_OFFSET: 10000,
    SE_FICT_OFFSET: 40,
    SE_FICT_OFFSET_1: 39,
    SE_FICT_MAX: 999,
    SE_NFICT_ELEM: 15,
    SE_COMET_OFFSET: 1000,
    SE_CUPIDO: 40,
    SE_HADES: 41,
    SE_ZEUS: 42,
    SE_KRONOS: 43,
    SE_APOLLON: 44,
    SE_ADMETOS: 45,
    SE_VULKANUS: 46,
    SE_POSEIDON: 47,
    SE_ISIS: 48,
    SE_NIBIRU: 49,
    SE_HARRINGTON: 50,
    SE_NEPTUNE_LEVERRIER: 51,
    SE_NEPTUNE_ADAMS: 52,
    SE_PLUTO_LOWELL: 53,
    SE_PLUTO_PICKERING: 54,
    SE_VULCAN: 55,
    SE_WHITE_MOON: 56,
    SE_PROSERPINA: 57,
    SE_WALDEMATH: 58,
    SE_FIXSTAR: -10,
    SE_ASC: 0,
    SE_MC: 1,
    SE_ARMC: 2,
    SE_VERTEX: 3,
    SE_EQUASC: 4,
    SE_COASC1: 5,
    SE_COASC2: 6,
    SE_POLASC: 7,
    SE_NASCMC: 8,
    SEFLG_JPLEPH: 1,
    SEFLG_SWIEPH: 2,
    SEFLG_MOSEPH: 4,
    SEFLG_HELCTR: 8,
    SEFLG_TRUEPOS: 16,
    SEFLG_J2000: 32,
    SEFLG_NONUT: 64,
    SEFLG_SPEED3: 128,
    SEFLG_SPEED: 256,
    SEFLG_NOGDEFL: 512,
    SEFLG_NOABERR: 1024,
    SEFLG_EQUATORIAL: 2 * 1024,
    SEFLG_XYZ: 4 * 1024,
    SEFLG_RADIANS: 8 * 1024,
    SEFLG_BARYCTR: 16 * 1024,
    SEFLG_TOPOCTR: 32 * 1024,
    SEFLG_TROPICAL: 0,
    SEFLG_SIDEREAL: 64 * 1024,
    SEFLG_ICRS: 128 * 1024,
    SEFLG_DPSIDEPS_1980: 256 * 1024,
    SEFLG_JPLHOR_APPROX: 512 * 1024,
    SEFLG_CENTER_BODY: 1024 * 1024,
    SE_SIDBITS: 256,
    SE_SIDBIT_ECL_T0: 256,
    SE_SIDBIT_SSY_PLANE: 512,
    SE_SIDBIT_USER_UT: 1024,
    SE_SIDBIT_ECL_DATE: 2048,
    SE_SIDBIT_NO_PREC_OFFSET: 4096,
    SE_SIDBIT_PREC_ORIG: 8192,
    SE_SIDM_FAGAN_BRADLEY: 0,
    SE_SIDM_LAHIRI: 1,
    SE_SIDM_DELUCE: 2,
    SE_SIDM_RAMAN: 3,
    SE_SIDM_USHASHASHI: 4,
    SE_SIDM_KRISHNAMURTI: 5,
    SE_SIDM_DJWHAL_KHUL: 6,
    SE_SIDM_YUKTESHWAR: 7,
    SE_SIDM_JN_BHASIN: 8,
    SE_SIDM_BABYL_KUGLER1: 9,
    SE_SIDM_BABYL_KUGLER2: 10,
    SE_SIDM_BABYL_KUGLER3: 11,
    SE_SIDM_BABYL_HUBER: 12,
    SE_SIDM_BABYL_ETPSC: 13,
    SE_SIDM_ALDEBARAN_15TAU: 14,
    SE_SIDM_HIPPARCHOS: 15,
    SE_SIDM_SASSANIAN: 16,
    SE_SIDM_GALCENT_0SAG: 17,
    SE_SIDM_J2000: 18,
    SE_SIDM_J1900: 19,
    SE_SIDM_B1950: 20,
    SE_SIDM_SURYASIDDHANTA: 21,
    SE_SIDM_SURYASIDDHANTA_MSUN: 22,
    SE_SIDM_ARYABHATA: 23,
    SE_SIDM_ARYABHATA_MSUN: 24,
    SE_SIDM_SS_REVATI: 25,
    SE_SIDM_SS_CITRA: 26,
    SE_SIDM_TRUE_CITRA: 27,
    SE_SIDM_TRUE_REVATI: 28,
    SE_SIDM_TRUE_PUSHYA: 29,
    SE_SIDM_GALCENT_RGILBRAND: 30,
    SE_SIDM_GALEQU_IAU1958: 31,
    SE_SIDM_GALEQU_TRUE: 32,
    SE_SIDM_GALEQU_MULA: 33,
    SE_SIDM_GALALIGN_MARDYKS: 34,
    SE_SIDM_TRUE_MULA: 35,
    SE_SIDM_GALCENT_MULA_WILHELM: 36,
    SE_SIDM_ARYABHATA_522: 37,
    SE_SIDM_BABYL_BRITTON: 38,
    SE_SIDM_TRUE_SHEORAN: 39,
    SE_SIDM_GALCENT_COCHRANE: 40,
    SE_SIDM_GALEQU_FIORENZA: 41,
    SE_SIDM_VALENS_MOON: 42,
    SE_SIDM_LAHIRI_1940: 43,
    SE_SIDM_LAHIRI_VP285: 44,
    SE_SIDM_KRISHNAMURTI_VP291: 45,
    SE_SIDM_LAHIRI_ICRC: 46,
    SE_SIDM_MANJULA: 43,
    SE_SIDM_USER: 255,
    SE_NSIDM_PREDEF: 47,
    SE_NODBIT_MEAN: 1,
    SE_NODBIT_OSCU: 2,
    SE_NODBIT_OSCU_BAR: 4,
    SE_NODBIT_FOPOINT: 256,
    SE_MAX_STNAME: 256,
    SE_ECL_CENTRAL: 1,
    SE_ECL_NONCENTRAL: 2,
    SE_ECL_TOTAL: 4,
    SE_ECL_ANNULAR: 8,
    SE_ECL_PARTIAL: 16,
    SE_ECL_ANNULAR_TOTAL: 32,
    SE_ECL_HYBRID: 32,
    SE_ECL_PENUMBRAL: 64,
    SE_ECL_VISIBLE: 128,
    SE_ECL_MAX_VISIBLE: 256,
    SE_ECL_1ST_VISIBLE: 512,
    SE_ECL_PARTBEG_VISIBLE: 512,
    SE_ECL_2ND_VISIBLE: 1024,
    SE_ECL_TOTBEG_VISIBLE: 1024,
    SE_ECL_3RD_VISIBLE: 2048,
    SE_ECL_TOTEND_VISIBLE: 2048,
    SE_ECL_4TH_VISIBLE: 4096,
    SE_ECL_PARTEND_VISIBLE: 4096,
    SE_ECL_PENUMBBEG_VISIBLE: 8192,
    SE_ECL_PENUMBEND_VISIBLE: 16384,
    SE_ECL_OCC_BEG_DAYLIGHT: 8192,
    SE_ECL_OCC_END_DAYLIGHT: 16384,
    SE_ECL_ONE_TRY: 32 * 1024,
    SE_CALC_RISE: 1,
    SE_CALC_SET: 2,
    SE_CALC_MTRANSIT: 4,
    SE_CALC_ITRANSIT: 8,
    SE_BIT_DISC_CENTER: 256,
    SE_BIT_DISC_BOTTOM: 8192,
    SE_BIT_GEOCTR_NO_ECL_LAT: 128,
    SE_BIT_NO_REFRACTION: 512,
    SE_BIT_CIVIL_TWILIGHT: 1024,
    SE_BIT_NAUTIC_TWILIGHT: 2048,
    SE_BIT_ASTRO_TWILIGHT: 4096,
    SE_BIT_FIXED_DISC_SIZE: 16384,
    SE_BIT_FORCE_SLOW_METHOD: 32768,
    SE_ECL2HOR: 0,
    SE_EQU2HOR: 1,
    SE_HOR2ECL: 0,
    SE_HOR2EQU: 1,
    SE_TRUE_TO_APP: 0,
    SE_APP_TO_TRUE: 1,
    SE_DE_NUMBER: 431,
    SE_SPLIT_DEG_ROUND_SEC: 1,
    SE_SPLIT_DEG_ROUND_MIN: 2,
    SE_SPLIT_DEG_ROUND_DEG: 4,
    SE_SPLIT_DEG_ZODIACAL: 8,
    SE_SPLIT_DEG_NAKSHATRA: 1024,
    SE_SPLIT_DEG_KEEP_SIGN: 16,
    SE_SPLIT_DEG_KEEP_DEG: 32,
    SE_HELIACAL_RISING: 1,
    SE_HELIACAL_SETTING: 2,
    SE_EVENING_FIRST: 3,
    SE_MORNING_LAST: 4,
    SE_ACRONYCHAL_RISING: 5,
    SE_ACRONYCHAL_SETTING: 6,
    SE_HELFLAG_LONG_SEARCH: 128,
    SE_HELFLAG_HIGH_PRECISION: 256,
    SE_HELFLAG_OPTICAL_PARAMS: 512,
    SE_HELFLAG_NO_DETAILS: 1024,
    SE_HELFLAG_SEARCH_1_PERIOD: 1 << 11,
    SE_HELFLAG_VISLIM_DARK: 1 << 12,
    SE_HELFLAG_VISLIM_NOMOON: 1 << 13,
    SE_HELFLAG_VISLIM_PHOTOPIC: 1 << 14,
    SE_HELFLAG_VISLIM_SCOTOPIC: 1 << 15,
    SE_HELFLAG_AV: 1 << 16,
    SE_HELFLAG_AVKIND_VR: 1 << 16,
    SE_HELFLAG_AVKIND_PTO: 1 << 17,
    SE_HELFLAG_AVKIND_MIN7: 1 << 18,
    SE_HELFLAG_AVKIND_MIN9: 1 << 19,
    TJD_INVALID: 99999999.0,
    SIMULATE_VICTORVB: 1,
    SE_HELIACAL_LONG_SEARCH: 128,
    SE_HELIACAL_HIGH_PRECISION: 256,
    SE_HELIACAL_OPTICAL_PARAMS: 512,
    SE_HELIACAL_NO_DETAILS: 1024,
    SE_HELIACAL_SEARCH_1_PERIOD: 1 << 11,
    SE_HELIACAL_VISLIM_DARK: 1 << 12,
    SE_HELIACAL_VISLIM_NOMOON: 1 << 13,
    SE_HELIACAL_VISLIM_PHOTOPIC: 1 << 14,
    SE_HELIACAL_AVKIND_VR: 1 << 15,
    SE_HELIACAL_AVKIND_PTO: 1 << 16,
    SE_HELIACAL_AVKIND_MIN7: 1 << 17,
    SE_HELIACAL_AVKIND_MIN9: 1 << 18,
    SE_PHOTOPIC_FLAG: 0,
    SE_SCOTOPIC_FLAG: 1,
    SE_MIXEDOPIC_FLAG: 2,
    SE_TIDAL_DE200: -23.8946,
    SE_TIDAL_DE403: -25.580,
    SE_TIDAL_DE404: -25.580,
    SE_TIDAL_DE405: -25.826,
    SE_TIDAL_DE406: -25.826,
    SE_TIDAL_DE421: -25.85,
    SE_TIDAL_DE422: -25.85,
    SE_TIDAL_DE430: -25.82,
    SE_TIDAL_DE431: -25.80,
    SE_TIDAL_DE441: -25.936,
    SE_TIDAL_26: -26.0,
    SE_TIDAL_STEPHENSON_2016: -25.85,
    SE_TIDAL_AUTOMATIC: 999999,
    SE_DELTAT_AUTOMATIC: -1E-10,
    SE_MODEL_DELTAT: 0,
    SE_MODEL_PREC_LONGTERM: 1,
    SE_MODEL_PREC_SHORTTERM: 2,
    SE_MODEL_NUT: 3,
    SE_MODEL_BIAS: 4,
    SE_MODEL_JPLHOR_MODE: 5,
    SE_MODEL_JPLHORA_MODE: 6,
    SE_MODEL_SIDT: 7,
    NSE_MODELS: 8,
    SEMOD_NPREC: 11,
    SEMOD_PREC_IAU_1976: 1,
    SEMOD_PREC_LASKAR_1986: 2,
    SEMOD_PREC_WILL_EPS_LASK: 3,
    SEMOD_PREC_WILLIAMS_1994: 4,
    SEMOD_PREC_SIMON_1994: 5,
    SEMOD_PREC_IAU_2000: 6,
    SEMOD_PREC_BRETAGNON_2003: 7,
    SEMOD_PREC_IAU_2006: 8,
    SEMOD_PREC_VONDRAK_2011: 9,
    SEMOD_PREC_OWEN_1990: 10,
    SEMOD_PREC_NEWCOMB: 11,
    SEMOD_NNUT: 5,
    SEMOD_NUT_IAU_1980: 1,
    SEMOD_NUT_IAU_CORR_1987: 2,
    SEMOD_NUT_IAU_2000A: 3,
    SEMOD_NUT_IAU_2000B: 4,
    SEMOD_NUT_WOOLARD: 5,
    SEMOD_NSIDT: 4,
    SEMOD_SIDT_IAU_1976: 1,
    SEMOD_SIDT_IAU_2006: 2,
    SEMOD_SIDT_IERS_CONV_2010: 3,
    SEMOD_SIDT_LONGTERM: 4,
    SEMOD_NBIAS: 3,
    SEMOD_BIAS_NONE: 1,
    SEMOD_BIAS_IAU2000: 2,
    SEMOD_BIAS_IAU2006: 3,
    SEMOD_NJPLHOR: 2,
    SEMOD_JPLHOR_LONG_AGREEMENT: 1,
    SEMOD_NJPLHORA: 3,
    SEMOD_JPLHORA_1: 1,
    SEMOD_JPLHORA_2: 2,
    SEMOD_JPLHORA_3: 3,
    SEMOD_NDELTAT: 5,
    SEMOD_DELTAT_STEPHENSON_MORRISON_1984: 1,
    SEMOD_DELTAT_STEPHENSON_1997: 2,
    SEMOD_DELTAT_STEPHENSON_MORRISON_2004: 3,
    SEMOD_DELTAT_ESPENAK_MEEUS_2006: 4,
    SEMOD_DELTAT_STEPHENSON_ETC_2016: 5,
    MY_FALSE: 0,
    TRUE: 1,
    FALSE: 0,
    OK: 0,
    ERR: -1,
    ODEGREE_STRING: "°",
    HUGE: 1.7E+308,
    M_PI: 3.14159265358979323846,
    AS_MAXCH: 256,
    DEG: 360000,
    DEG7_30: 2700000,
    CS2DEG: 1.0 / 360000.0,
    O_BINARY: 0,
    OPEN_MODE: 0o666,
    DIR_GLUE: "\\",
    PATH_SEPARATOR: ";",
};
export const Signatures = [
    {
        "returnType": "int32",
        "name": "swe_heliacal_ut",
        "args": [
            {
                "type": "double",
                "name": "tjdstart_ut"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "datm"
            },
            {
                "type": "double*",
                "name": "dobs"
            },
            {
                "type": "char*",
                "name": "ObjectName"
            },
            {
                "type": "int32",
                "name": "TypeEvent"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "dret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_heliacal_pheno_ut",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "datm"
            },
            {
                "type": "double*",
                "name": "dobs"
            },
            {
                "type": "char*",
                "name": "ObjectName"
            },
            {
                "type": "int32",
                "name": "TypeEvent"
            },
            {
                "type": "int32",
                "name": "helflag"
            },
            {
                "type": "double*",
                "name": "darr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_vis_limit_mag",
        "args": [
            {
                "type": "double",
                "name": "tjdut"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "datm"
            },
            {
                "type": "double*",
                "name": "dobs"
            },
            {
                "type": "char*",
                "name": "ObjectName"
            },
            {
                "type": "int32",
                "name": "helflag"
            },
            {
                "type": "double*",
                "name": "dret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_heliacal_angle",
        "args": [
            {
                "type": "double",
                "name": "tjdut"
            },
            {
                "type": "double*",
                "name": "dgeo"
            },
            {
                "type": "double*",
                "name": "datm"
            },
            {
                "type": "double*",
                "name": "dobs"
            },
            {
                "type": "int32",
                "name": "helflag"
            },
            {
                "type": "double",
                "name": "mag"
            },
            {
                "type": "double",
                "name": "azi_obj"
            },
            {
                "type": "double",
                "name": "azi_sun"
            },
            {
                "type": "double",
                "name": "azi_moon"
            },
            {
                "type": "double",
                "name": "alt_moon"
            },
            {
                "type": "double*",
                "name": "dret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_topo_arcus_visionis",
        "args": [
            {
                "type": "double",
                "name": "tjdut"
            },
            {
                "type": "double*",
                "name": "dgeo"
            },
            {
                "type": "double*",
                "name": "datm"
            },
            {
                "type": "double*",
                "name": "dobs"
            },
            {
                "type": "int32",
                "name": "helflag"
            },
            {
                "type": "double",
                "name": "mag"
            },
            {
                "type": "double",
                "name": "azi_obj"
            },
            {
                "type": "double",
                "name": "alt_obj"
            },
            {
                "type": "double",
                "name": "azi_sun"
            },
            {
                "type": "double",
                "name": "azi_moon"
            },
            {
                "type": "double",
                "name": "alt_moon"
            },
            {
                "type": "double*",
                "name": "dret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_astro_models",
        "args": [
            {
                "type": "char*",
                "name": "samod"
            },
            {
                "type": "int32",
                "name": "iflag"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_get_astro_models",
        "args": [
            {
                "type": "char*",
                "name": "samod"
            },
            {
                "type": "char*",
                "name": "sdet"
            },
            {
                "type": "int32",
                "name": "iflag"
            }
        ]
    },
    {
        "returnType": "char *",
        "name": "swe_version",
        "args": [
            {
                "type": "char*",
                "name": ""
            }
        ]
    },
    {
        "returnType": "char *",
        "name": "swe_get_library_path",
        "args": [
            {
                "type": "char*",
                "name": ""
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_calc",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xx"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_calc_ut",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xx"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_calc_pctr",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iplctr"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xxret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_solcross",
        "args": [
            {
                "type": "double",
                "name": "x2cross"
            },
            {
                "type": "double",
                "name": "jd_et"
            },
            {
                "type": "int32",
                "name": "flag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_solcross_ut",
        "args": [
            {
                "type": "double",
                "name": "x2cross"
            },
            {
                "type": "double",
                "name": "jd_ut"
            },
            {
                "type": "int32",
                "name": "flag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_mooncross",
        "args": [
            {
                "type": "double",
                "name": "x2cross"
            },
            {
                "type": "double",
                "name": "jd_et"
            },
            {
                "type": "int32",
                "name": "flag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_mooncross_ut",
        "args": [
            {
                "type": "double",
                "name": "x2cross"
            },
            {
                "type": "double",
                "name": "jd_ut"
            },
            {
                "type": "int32",
                "name": "flag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_mooncross_node",
        "args": [
            {
                "type": "double",
                "name": "jd_et"
            },
            {
                "type": "int32",
                "name": "flag"
            },
            {
                "type": "double*",
                "name": "xlon"
            },
            {
                "type": "double*",
                "name": "xlat"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_mooncross_node_ut",
        "args": [
            {
                "type": "double",
                "name": "jd_ut"
            },
            {
                "type": "int32",
                "name": "flag"
            },
            {
                "type": "double*",
                "name": "xlon"
            },
            {
                "type": "double*",
                "name": "xlat"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_helio_cross",
        "args": [
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "double",
                "name": "x2cross"
            },
            {
                "type": "double",
                "name": "jd_et"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "int32",
                "name": "dir"
            },
            {
                "type": "double*",
                "name": "jd_cross"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_helio_cross_ut",
        "args": [
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "double",
                "name": "x2cross"
            },
            {
                "type": "double",
                "name": "jd_ut"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "int32",
                "name": "dir"
            },
            {
                "type": "double*",
                "name": "jd_cross"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_fixstar",
        "args": [
            {
                "type": "char*",
                "name": "star"
            },
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xx"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_fixstar_ut",
        "args": [
            {
                "type": "char*",
                "name": "star"
            },
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xx"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_fixstar_mag",
        "args": [
            {
                "type": "char*",
                "name": "star"
            },
            {
                "type": "double*",
                "name": "mag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_fixstar2",
        "args": [
            {
                "type": "char*",
                "name": "star"
            },
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xx"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_fixstar2_ut",
        "args": [
            {
                "type": "char*",
                "name": "star"
            },
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "xx"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_fixstar2_mag",
        "args": [
            {
                "type": "char*",
                "name": "star"
            },
            {
                "type": "double*",
                "name": "mag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_close",
        "args": [
            {
                "type": "void",
                "name": ""
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_ephe_path",
        "args": [
            {
                "type": "const char*",
                "name": "path"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_jpl_file",
        "args": [
            {
                "type": "const char*",
                "name": "fname"
            }
        ]
    },
    {
        "returnType": "char *",
        "name": "swe_get_planet_name",
        "args": [
            {
                "type": "int",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "spname"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_topo",
        "args": [
            {
                "type": "double",
                "name": "geolon"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "geoalt"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_sid_mode",
        "args": [
            {
                "type": "int32",
                "name": "sid_mode"
            },
            {
                "type": "double",
                "name": "t0"
            },
            {
                "type": "double",
                "name": "ayan_t0"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_get_ayanamsa_ex",
        "args": [
            {
                "type": "double",
                "name": "tjd_et"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "daya"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_get_ayanamsa_ex_ut",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "daya"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_get_ayanamsa",
        "args": [
            {
                "type": "double",
                "name": "tjd_et"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_get_ayanamsa_ut",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            }
        ]
    },
    {
        "returnType": "const char *",
        "name": "swe_get_ayanamsa_name",
        "args": [
            {
                "type": "int32",
                "name": "isidmode"
            }
        ]
    },
    {
        "returnType": "const char *",
        "name": "swe_get_current_file_data",
        "args": [
            {
                "type": "int",
                "name": "ifno"
            },
            {
                "type": "double*",
                "name": "tfstart"
            },
            {
                "type": "double*",
                "name": "tfend"
            },
            {
                "type": "int*",
                "name": "denum"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_timeout",
        "args": [
            {
                "type": "int32",
                "name": "tsec"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_date_conversion",
        "args": [
            {
                "type": "int",
                "name": "y"
            },
            {
                "type": "int",
                "name": "m"
            },
            {
                "type": "int",
                "name": "d"
            },
            {
                "type": "/*",
                "name": "year"
            },
            {
                "type": "month",
                "name": ""
            },
            {
                "type": "day /\n        double*",
                "name": "utime"
            },
            {
                "type": "/ universal time in hours (decimal) /\n        char*",
                "name": "c"
            },
            {
                "type": "/ calendar g[regorian]|j[ulian] /\n        double*",
                "name": "tjd"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_julday",
        "args": [
            {
                "type": "int",
                "name": "year"
            },
            {
                "type": "int",
                "name": "month"
            },
            {
                "type": "int",
                "name": "day"
            },
            {
                "type": "double",
                "name": "hour"
            },
            {
                "type": "int",
                "name": "gregflag"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_revjul",
        "args": [
            {
                "type": "double",
                "name": "jd"
            },
            {
                "type": "int",
                "name": "gregflag"
            },
            {
                "type": "int*",
                "name": "jyear"
            },
            {
                "type": "int*",
                "name": "jmon"
            },
            {
                "type": "int*",
                "name": "jday"
            },
            {
                "type": "double*",
                "name": "jut"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_utc_to_jd",
        "args": [
            {
                "type": "int32",
                "name": "iyear"
            },
            {
                "type": "int32",
                "name": "imonth"
            },
            {
                "type": "int32",
                "name": "iday"
            },
            {
                "type": "int32",
                "name": "ihour"
            },
            {
                "type": "int32",
                "name": "imin"
            },
            {
                "type": "double",
                "name": "dsec"
            },
            {
                "type": "int32",
                "name": "gregflag"
            },
            {
                "type": "double*",
                "name": "dret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_jdet_to_utc",
        "args": [
            {
                "type": "double",
                "name": "tjd_et"
            },
            {
                "type": "int32",
                "name": "gregflag"
            },
            {
                "type": "int32*",
                "name": "iyear"
            },
            {
                "type": "int32*",
                "name": "imonth"
            },
            {
                "type": "int32*",
                "name": "iday"
            },
            {
                "type": "int32*",
                "name": "ihour"
            },
            {
                "type": "int32*",
                "name": "imin"
            },
            {
                "type": "double*",
                "name": "dsec"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_jdut1_to_utc",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "gregflag"
            },
            {
                "type": "int32*",
                "name": "iyear"
            },
            {
                "type": "int32*",
                "name": "imonth"
            },
            {
                "type": "int32*",
                "name": "iday"
            },
            {
                "type": "int32*",
                "name": "ihour"
            },
            {
                "type": "int32*",
                "name": "imin"
            },
            {
                "type": "double*",
                "name": "dsec"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_utc_time_zone",
        "args": [
            {
                "type": "int32",
                "name": "iyear"
            },
            {
                "type": "int32",
                "name": "imonth"
            },
            {
                "type": "int32",
                "name": "iday"
            },
            {
                "type": "int32",
                "name": "ihour"
            },
            {
                "type": "int32",
                "name": "imin"
            },
            {
                "type": "double",
                "name": "dsec"
            },
            {
                "type": "double",
                "name": "d_timezone"
            },
            {
                "type": "int32*",
                "name": "iyear_out"
            },
            {
                "type": "int32*",
                "name": "imonth_out"
            },
            {
                "type": "int32*",
                "name": "iday_out"
            },
            {
                "type": "int32*",
                "name": "ihour_out"
            },
            {
                "type": "int32*",
                "name": "imin_out"
            },
            {
                "type": "double*",
                "name": "dsec_out"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_houses",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "geolon"
            },
            {
                "type": "int",
                "name": "hsys"
            },
            {
                "type": "double*",
                "name": "cusps"
            },
            {
                "type": "double*",
                "name": "ascmc"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_houses_ex",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "geolon"
            },
            {
                "type": "int",
                "name": "hsys"
            },
            {
                "type": "double*",
                "name": "cusps"
            },
            {
                "type": "double*",
                "name": "ascmc"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_houses_ex2",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "geolon"
            },
            {
                "type": "int",
                "name": "hsys"
            },
            {
                "type": "double*",
                "name": "cusps"
            },
            {
                "type": "double*",
                "name": "ascmc"
            },
            {
                "type": "double*",
                "name": "cusp_speed"
            },
            {
                "type": "double*",
                "name": "ascmc_speed"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_houses_armc",
        "args": [
            {
                "type": "double",
                "name": "armc"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "eps"
            },
            {
                "type": "int",
                "name": "hsys"
            },
            {
                "type": "double*",
                "name": "cusps"
            },
            {
                "type": "double*",
                "name": "ascmc"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_houses_armc_ex2",
        "args": [
            {
                "type": "double",
                "name": "armc"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "eps"
            },
            {
                "type": "int",
                "name": "hsys"
            },
            {
                "type": "double*",
                "name": "cusps"
            },
            {
                "type": "double*",
                "name": "ascmc"
            },
            {
                "type": "double*",
                "name": "cusp_speed"
            },
            {
                "type": "double*",
                "name": "ascmc_speed"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_house_pos",
        "args": [
            {
                "type": "double",
                "name": "armc"
            },
            {
                "type": "double",
                "name": "geolat"
            },
            {
                "type": "double",
                "name": "eps"
            },
            {
                "type": "int",
                "name": "hsys"
            },
            {
                "type": "double*",
                "name": "xpin"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "const char *",
        "name": "swe_house_name",
        "args": [
            {
                "type": "int",
                "name": "hsys"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_gauquelin_sector",
        "args": [
            {
                "type": "double",
                "name": "t_ut"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "starname"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "int32",
                "name": "imeth"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double",
                "name": "atpress"
            },
            {
                "type": "double",
                "name": "attemp"
            },
            {
                "type": "double*",
                "name": "dgsect"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_sol_eclipse_where",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lun_occult_where",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "starname"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_sol_eclipse_how",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_sol_eclipse_when_loc",
        "args": [
            {
                "type": "double",
                "name": "tjd_start"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "int32",
                "name": "backward"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lun_occult_when_loc",
        "args": [
            {
                "type": "double",
                "name": "tjd_start"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "starname"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "int32",
                "name": "backward"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_sol_eclipse_when_glob",
        "args": [
            {
                "type": "double",
                "name": "tjd_start"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "int32",
                "name": "ifltype"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "int32",
                "name": "backward"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lun_occult_when_glob",
        "args": [
            {
                "type": "double",
                "name": "tjd_start"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "starname"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "int32",
                "name": "ifltype"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "int32",
                "name": "backward"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lun_eclipse_how",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lun_eclipse_when",
        "args": [
            {
                "type": "double",
                "name": "tjd_start"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "int32",
                "name": "ifltype"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "int32",
                "name": "backward"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lun_eclipse_when_loc",
        "args": [
            {
                "type": "double",
                "name": "tjd_start"
            },
            {
                "type": "int32",
                "name": "ifl"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "int32",
                "name": "backward"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_pheno",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_pheno_ut",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "attr"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_refrac",
        "args": [
            {
                "type": "double",
                "name": "inalt"
            },
            {
                "type": "double",
                "name": "atpress"
            },
            {
                "type": "double",
                "name": "attemp"
            },
            {
                "type": "int32",
                "name": "calc_flag"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_refrac_extended",
        "args": [
            {
                "type": "double",
                "name": "inalt"
            },
            {
                "type": "double",
                "name": "geoalt"
            },
            {
                "type": "double",
                "name": "atpress"
            },
            {
                "type": "double",
                "name": "attemp"
            },
            {
                "type": "double",
                "name": "lapse_rate"
            },
            {
                "type": "int32",
                "name": "calc_flag"
            },
            {
                "type": "double*",
                "name": "dret"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_lapse_rate",
        "args": [
            {
                "type": "double",
                "name": "lapse_rate"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_azalt",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "calc_flag"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double",
                "name": "atpress"
            },
            {
                "type": "double",
                "name": "attemp"
            },
            {
                "type": "double*",
                "name": "xin"
            },
            {
                "type": "double*",
                "name": "xaz"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_azalt_rev",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "calc_flag"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double*",
                "name": "xin"
            },
            {
                "type": "double*",
                "name": "xout"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_rise_trans_true_hor",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "starname"
            },
            {
                "type": "int32",
                "name": "epheflag"
            },
            {
                "type": "int32",
                "name": "rsmi"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double",
                "name": "atpress"
            },
            {
                "type": "double",
                "name": "attemp"
            },
            {
                "type": "double",
                "name": "horhgt"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_rise_trans",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "char*",
                "name": "starname"
            },
            {
                "type": "int32",
                "name": "epheflag"
            },
            {
                "type": "int32",
                "name": "rsmi"
            },
            {
                "type": "double*",
                "name": "geopos"
            },
            {
                "type": "double",
                "name": "atpress"
            },
            {
                "type": "double",
                "name": "attemp"
            },
            {
                "type": "double*",
                "name": "tret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_nod_aps",
        "args": [
            {
                "type": "double",
                "name": "tjd_et"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "int32",
                "name": "method"
            },
            {
                "type": "double*",
                "name": "xnasc"
            },
            {
                "type": "double*",
                "name": "xndsc"
            },
            {
                "type": "double*",
                "name": "xperi"
            },
            {
                "type": "double*",
                "name": "xaphe"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_nod_aps_ut",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "int32",
                "name": "method"
            },
            {
                "type": "double*",
                "name": "xnasc"
            },
            {
                "type": "double*",
                "name": "xndsc"
            },
            {
                "type": "double*",
                "name": "xperi"
            },
            {
                "type": "double*",
                "name": "xaphe"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_get_orbital_elements",
        "args": [
            {
                "type": "double",
                "name": "tjd_et"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "dret"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_orbit_max_min_true_distance",
        "args": [
            {
                "type": "double",
                "name": "tjd_et"
            },
            {
                "type": "int32",
                "name": "ipl"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "double*",
                "name": "dmax"
            },
            {
                "type": "double*",
                "name": "dmin"
            },
            {
                "type": "double*",
                "name": "dtrue"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_deltat",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_deltat_ex",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "int32",
                "name": "iflag"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_time_equ",
        "args": [
            {
                "type": "double",
                "name": "tjd"
            },
            {
                "type": "double*",
                "name": "te"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lmt_to_lat",
        "args": [
            {
                "type": "double",
                "name": "tjd_lmt"
            },
            {
                "type": "double",
                "name": "geolon"
            },
            {
                "type": "double*",
                "name": "tjd_lat"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_lat_to_lmt",
        "args": [
            {
                "type": "double",
                "name": "tjd_lat"
            },
            {
                "type": "double",
                "name": "geolon"
            },
            {
                "type": "double*",
                "name": "tjd_lmt"
            },
            {
                "type": "char*",
                "name": "serr"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_sidtime0",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            },
            {
                "type": "double",
                "name": "eps"
            },
            {
                "type": "double",
                "name": "nut"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_sidtime",
        "args": [
            {
                "type": "double",
                "name": "tjd_ut"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_interpolate_nut",
        "args": [
            {
                "type": "AS_BOOL",
                "name": "do_interpolate"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_cotrans",
        "args": [
            {
                "type": "double*",
                "name": "xpo"
            },
            {
                "type": "double*",
                "name": "xpn"
            },
            {
                "type": "double",
                "name": "eps"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_cotrans_sp",
        "args": [
            {
                "type": "double*",
                "name": "xpo"
            },
            {
                "type": "double*",
                "name": "xpn"
            },
            {
                "type": "double",
                "name": "eps"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_get_tid_acc",
        "args": [
            {
                "type": "void",
                "name": ""
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_tid_acc",
        "args": [
            {
                "type": "double",
                "name": "t_acc"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_set_delta_t_userdef",
        "args": [
            {
                "type": "double",
                "name": "dt"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_degnorm",
        "args": [
            {
                "type": "double",
                "name": "x"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_radnorm",
        "args": [
            {
                "type": "double",
                "name": "x"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_rad_midp",
        "args": [
            {
                "type": "double",
                "name": "x1"
            },
            {
                "type": "double",
                "name": "x0"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_deg_midp",
        "args": [
            {
                "type": "double",
                "name": "x1"
            },
            {
                "type": "double",
                "name": "x0"
            }
        ]
    },
    {
        "returnType": "void",
        "name": "swe_split_deg",
        "args": [
            {
                "type": "double",
                "name": "ddeg"
            },
            {
                "type": "int32",
                "name": "roundflag"
            },
            {
                "type": "int32*",
                "name": "ideg"
            },
            {
                "type": "int32*",
                "name": "imin"
            },
            {
                "type": "int32*",
                "name": "isec"
            },
            {
                "type": "double*",
                "name": "dsecfr"
            },
            {
                "type": "int32*",
                "name": "isgn"
            }
        ]
    },
    {
        "returnType": "centisec",
        "name": "swe_csnorm",
        "args": [
            {
                "type": "centisec",
                "name": "p"
            }
        ]
    },
    {
        "returnType": "centisec",
        "name": "swe_difcsn",
        "args": [
            {
                "type": "centisec",
                "name": "p1"
            },
            {
                "type": "centisec",
                "name": "p2"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_difdegn",
        "args": [
            {
                "type": "double",
                "name": "p1"
            },
            {
                "type": "double",
                "name": "p2"
            }
        ]
    },
    {
        "returnType": "centisec",
        "name": "swe_difcs2n",
        "args": [
            {
                "type": "centisec",
                "name": "p1"
            },
            {
                "type": "centisec",
                "name": "p2"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_difdeg2n",
        "args": [
            {
                "type": "double",
                "name": "p1"
            },
            {
                "type": "double",
                "name": "p2"
            }
        ]
    },
    {
        "returnType": "double",
        "name": "swe_difrad2n",
        "args": [
            {
                "type": "double",
                "name": "p1"
            },
            {
                "type": "double",
                "name": "p2"
            }
        ]
    },
    {
        "returnType": "centisec",
        "name": "swe_csroundsec",
        "args": [
            {
                "type": "centisec",
                "name": "x"
            }
        ]
    },
    {
        "returnType": "int32",
        "name": "swe_d2l",
        "args": [
            {
                "type": "double",
                "name": "x"
            }
        ]
    },
    {
        "returnType": "int",
        "name": "swe_day_of_week",
        "args": [
            {
                "type": "double",
                "name": "jd"
            }
        ]
    },
    {
        "returnType": "char *",
        "name": "swe_cs2timestr",
        "args": [
            {
                "type": "CSEC",
                "name": "t"
            },
            {
                "type": "int",
                "name": "sep"
            },
            {
                "type": "AS_BOOL",
                "name": "suppressZero"
            },
            {
                "type": "char*",
                "name": "a"
            }
        ]
    },
    {
        "returnType": "char *",
        "name": "swe_cs2lonlatstr",
        "args": [
            {
                "type": "CSEC",
                "name": "t"
            },
            {
                "type": "char",
                "name": "pchar"
            },
            {
                "type": "char",
                "name": "mchar"
            },
            {
                "type": "char*",
                "name": "s"
            }
        ]
    },
    {
        "returnType": "char *",
        "name": "swe_cs2degstr",
        "args": [
            {
                "type": "CSEC",
                "name": "t"
            },
            {
                "type": "char*",
                "name": "a"
            }
        ]
    }
];
