// @fusionstrings/swiss-eph/script/heap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WasmHeap = void 0;
/**
 * WebAssembly Memory Heap Manager.
 *
 * Provides typed access to the WASM linear memory, handling
 * allocation, deallocation, and data marshaling between JavaScript
 * and WebAssembly.
 */
class WasmHeap {
    constructor(memory, exports) {
        Object.defineProperty(this, "memory", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: memory
        });
        Object.defineProperty(this, "exports", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: exports
        });
        Object.defineProperty(this, "allocated", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Map()
        });
    }
    alloc(size) {
        let ptr;
        if (this.exports.malloc) {
            ptr = this.exports.malloc(size);
        }
        else if (this.exports.__wbindgen_malloc) {
            ptr = this.exports.__wbindgen_malloc(size);
        }
        else {
            throw new Error(`No malloc found in WASM exports. Available: ${Object.keys(this.exports).join(", ")}`);
        }
        this.allocated.set(ptr, size);
        return ptr;
    }
    free(ptr) {
        if (this.exports.free) {
            this.exports.free(ptr);
        }
        else if (this.exports.__wbindgen_free) {
            const size = this.allocated.get(ptr) || 0;
            this.exports.__wbindgen_free(ptr, size, 8);
        }
        this.allocated.delete(ptr);
    }
    getU8(ptr, length) {
        return new Uint8Array(this.memory.buffer, ptr, length);
    }
    setU8(ptr, data) {
        new Uint8Array(this.memory.buffer, ptr, data.length).set(data);
    }
    getF64(ptr, length) {
        return new Float64Array(this.memory.buffer, ptr, length);
    }
    getString(ptr) {
        const buffer = new Uint8Array(this.memory.buffer);
        let end = ptr;
        while (buffer[end] !== 0)
            end++;
        return new TextDecoder().decode(buffer.subarray(ptr, end));
    }
    getI32(ptr) {
        return new DataView(this.memory.buffer).getInt32(ptr, true);
    }
    putString(str) {
        const bytes = new TextEncoder().encode(str + "\0");
        const ptr = this.alloc(bytes.length);
        this.setU8(ptr, bytes);
        return ptr;
    }
}
exports.WasmHeap = WasmHeap;
