// @fusionstrings/swiss-eph/esm/lib/swiss_eph.internal.js
// @generated file from wasmbuild -- do not edit
// @ts-nocheck: generated
// deno-lint-ignore-file
// deno-fmt-ignore-file
let wasm;
export function __wbg_set_wasm(val) {
    wasm = val;
}
function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return decodeText(ptr, len);
}
let cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}
function passStringToWasm0(arg, malloc, realloc) {
    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }
    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;
    const mem = getUint8ArrayMemory0();
    let offset = 0;
    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F)
            break;
        mem[ptr + offset] = code;
    }
    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = cachedTextEncoder.encodeInto(arg, view);
        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }
    WASM_VECTOR_LEN = offset;
    return ptr;
}
function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_externrefs.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}
let cachedTextDecoder = new TextDecoder("utf-8", {
    ignoreBOM: true,
    fatal: true,
});
cachedTextDecoder.decode();
const MAX_SAFARI_DECODE_BYTES = 2146435072;
let numBytesDecoded = 0;
function decodeText(ptr, len) {
    numBytesDecoded += len;
    if (numBytesDecoded >= MAX_SAFARI_DECODE_BYTES) {
        cachedTextDecoder = new TextDecoder("utf-8", {
            ignoreBOM: true,
            fatal: true,
        });
        cachedTextDecoder.decode();
        numBytesDecoded = len;
    }
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}
const cachedTextEncoder = new TextEncoder();
if (!("encodeInto" in cachedTextEncoder)) {
    cachedTextEncoder.encodeInto = function (arg, view) {
        const buf = cachedTextEncoder.encode(arg);
        view.set(buf);
        return {
            read: arg.length,
            written: buf.length,
        };
    };
}
let WASM_VECTOR_LEN = 0;
const PositionFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_position_free(ptr >>> 0, 1));
const SwissEphErrorFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_swissepherror_free(ptr >>> 0, 1));
/**
 * Planetary position result
 */
export class Position {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(Position.prototype);
        obj.__wbg_ptr = ptr;
        PositionFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        PositionFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_position_free(ptr, 0);
    }
    /**
     * Ecliptic longitude in degrees
     * @returns {number}
     */
    get longitude() {
        const ret = wasm.__wbg_get_position_longitude(this.__wbg_ptr);
        return ret;
    }
    /**
     * Ecliptic longitude in degrees
     * @param {number} arg0
     */
    set longitude(arg0) {
        wasm.__wbg_set_position_longitude(this.__wbg_ptr, arg0);
    }
    /**
     * Ecliptic latitude in degrees
     * @returns {number}
     */
    get latitude() {
        const ret = wasm.__wbg_get_position_latitude(this.__wbg_ptr);
        return ret;
    }
    /**
     * Ecliptic latitude in degrees
     * @param {number} arg0
     */
    set latitude(arg0) {
        wasm.__wbg_set_position_latitude(this.__wbg_ptr, arg0);
    }
    /**
     * Distance (AU for planets, Earth radii for Moon)
     * @returns {number}
     */
    get distance() {
        const ret = wasm.__wbg_get_position_distance(this.__wbg_ptr);
        return ret;
    }
    /**
     * Distance (AU for planets, Earth radii for Moon)
     * @param {number} arg0
     */
    set distance(arg0) {
        wasm.__wbg_set_position_distance(this.__wbg_ptr, arg0);
    }
    /**
     * Longitude speed (degrees/day)
     * @returns {number}
     */
    get longitude_speed() {
        const ret = wasm.__wbg_get_position_longitude_speed(this.__wbg_ptr);
        return ret;
    }
    /**
     * Longitude speed (degrees/day)
     * @param {number} arg0
     */
    set longitude_speed(arg0) {
        wasm.__wbg_set_position_longitude_speed(this.__wbg_ptr, arg0);
    }
    /**
     * Latitude speed (degrees/day)
     * @returns {number}
     */
    get latitude_speed() {
        const ret = wasm.__wbg_get_position_latitude_speed(this.__wbg_ptr);
        return ret;
    }
    /**
     * Latitude speed (degrees/day)
     * @param {number} arg0
     */
    set latitude_speed(arg0) {
        wasm.__wbg_set_position_latitude_speed(this.__wbg_ptr, arg0);
    }
    /**
     * Distance speed (AU/day)
     * @returns {number}
     */
    get distance_speed() {
        const ret = wasm.__wbg_get_position_distance_speed(this.__wbg_ptr);
        return ret;
    }
    /**
     * Distance speed (AU/day)
     * @param {number} arg0
     */
    set distance_speed(arg0) {
        wasm.__wbg_set_position_distance_speed(this.__wbg_ptr, arg0);
    }
}
if (Symbol.dispose) {
    Position.prototype[Symbol.dispose] = Position.prototype.free;
}
/**
 * Error returned by Swiss Ephemeris calculations
 */
export class SwissEphError {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(SwissEphError.prototype);
        obj.__wbg_ptr = ptr;
        SwissEphErrorFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        SwissEphErrorFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_swissepherror_free(ptr, 0);
    }
    /**
     * Error message from the library
     * @returns {string}
     */
    get message() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.__wbg_get_swissepherror_message(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        }
        finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Error message from the library
     * @param {string} arg0
     */
    set message(arg0) {
        const ptr0 = passStringToWasm0(arg0, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.__wbg_set_swissepherror_message(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Return code
     * @returns {number}
     */
    get code() {
        const ret = wasm.__wbg_get_swissepherror_code(this.__wbg_ptr);
        return ret;
    }
    /**
     * Return code
     * @param {number} arg0
     */
    set code(arg0) {
        wasm.__wbg_set_swissepherror_code(this.__wbg_ptr, arg0);
    }
}
if (Symbol.dispose) {
    SwissEphError.prototype[Symbol.dispose] = SwissEphError.prototype.free;
}
/**
 * Calculate planetary position using UT (Universal Time)
 * @param {number} jd_ut
 * @param {number} planet
 * @param {number} flags
 * @returns {Position}
 */
export function calc_ut(jd_ut, planet, flags) {
    const ret = wasm.calc_ut(jd_ut, planet, flags);
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return Position.__wrap(ret[0]);
}
/**
 * Set the ephemeris path
 * @param {string} path
 */
export function set_ephe_path(path) {
    const ptr0 = passStringToWasm0(path, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len0 = WASM_VECTOR_LEN;
    wasm.set_ephe_path(ptr0, len0);
}
/**
 * Get Swiss Ephemeris version
 * @returns {string}
 */
export function version() {
    let deferred1_0;
    let deferred1_1;
    try {
        const ret = wasm.version();
        deferred1_0 = ret[0];
        deferred1_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    }
    finally {
        wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
    }
}
export function __wbg___wbindgen_throw_dd24417ed36fc46e(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
}
export function __wbg_swissepherror_new(arg0) {
    const ret = SwissEphError.__wrap(arg0);
    return ret;
}
export function __wbindgen_init_externref_table() {
    const table = wasm.__wbindgen_externrefs;
    const offset = table.grow(4);
    table.set(0, undefined);
    table.set(offset + 0, undefined);
    table.set(offset + 1, null);
    table.set(offset + 2, true);
    table.set(offset + 3, false);
}
